pkg/controller/volume/attachdetach/attach_detach_controller.go

```diff
@@ -390,12 +388,8 @@ func (adc *attachDetachController) populateActualStateOfWorld(logger klog.Logger
 				logger.Error(err, "Failed to mark the volume as attached")
 				continue
 			}
-			inUse := volumesInUse.Has(uniqueName)
-			err = adc.actualStateOfWorld.SetVolumeMountedByNode(logger, uniqueName, nodeName, inUse)
-			if err != nil {
-				logger.Error(err, "Failed to set volume mounted by node")
-			}
 		}
+		adc.actualStateOfWorld.SetVolumesMountedByNode(logger, node.Status.VolumesInUse, nodeName)
```

Comment:
Yes, this is intended. We strictly record more information than before. We can still know whether each attached volume is inUse.

Reply:
But A/D controller should worry about attachable volumes only. If I am reading `SetVolumeMountedByNode` right, it calls `getNodeAndVolume()` and that will return an error if the volume is not marked as attached
https://github.com/kubernetes/kubernetes/blob/4a4940694f40e8d74b73e8dfdfcec9c953e3fbfe/pkg/controller/volume/attachdetach/cache/actual_state_of_world.go#L463-L465

Reply:
It will just flood log with unnecessary errors.

Reply:
I think it makes sense to have a set of attached volumes and filters `volumesInUse` with it. It won't be exactly O(n), more like O(n*log(n)) or what `set.Has` overhead is.

Reply:
> It will just flood log with unnecessary errors.

Yes, so this PR should remove those unnecessary errors.

> But A/D controller should worry about attachable volumes only.

I agree, but finding out which volume is attachable can be hard. It may involve parsing the name and querying each plugin. And the extra cost here is just a larger `Set`. The memory under `UniqueVolumeName` should be shared with informer.

> have a set of attached volumes and filters volumesInUse with it

No, I intend to store in-use state for not-yet-attached volumes. So that I can (in a future PR) wait for the volume to become in-use before initiate the attaching.

Reply:
To elaborate more on why we should wait for in-use before attaching, consider this sequence of event in the current version:
1. A/D controller observed the new Pod, then initiate attaching
   A/D controller state: attached=true, inUse=true
2. A/D controller received an unrelated node update
   A/D controller state: attached=true, inUse=false
3. kubelet observed the new Pod, start to mount the file system in the volume.
4. the Pod is deleted
5. A/D controller observed the Pod was removed, then initiate detaching, since it thinks the volume is not in use
   A/D controller state: attached=false, inUse=undefined
6. A/D controller observed the node updated by kubelet, with new volume in-use, but this is too late.

So theoretically, the volume may detached before the kubelet has done the cleanup.

Reply:
It would be awesome if we could fix this issue! Thanks for working on it (and sorry about long reviews, this needs a large chunk of time to focus on the current code)